<center> <img src ="https://i.postimg.cc/1X8H7YYt/BITS-Logo.png" width = "400" alt="BITS Pilani Logo" /> </center>

<font color='green'> <h1> <center> Basics of NumPy Arrays </center> </h1> </font>

In [1]:
import numpy
numpy.__version__

'2.0.2'

NumPy (Numeric Python) provides an efficient way of storage and accessing the elements in numeric formats. They provide an easy to use interface by which the elements can be stored in a dense data structure and can be efficiently accessed whenever required.

Note - numpy is already installed when you are using Anaconda navigator.

In [2]:
import numpy as np

Usually alias 'np' is used to refer the numpy library.

<b> Python Arrays<b>

Python has arrays that stores the data in efficient manner which is having a fixed data type. The built-in array module helps in doing so.

In [3]:
import array
my_array = array.array('i', [1, 2, 3, 4])

print("type   : ", type(my_array))
my_array

type   :  <class 'array.array'>


array('i', [1, 2, 3, 4])

The above code snippet creates a array holding all the elements which are of same type i.e. integer. 'i', designates interger.

In [4]:
my_float_array = array.array('f', [1.1, 2.3, 4.5])
print("type   : ", type(my_float_array))
my_float_array

type   :  <class 'array.array'>


array('f', [1.100000023841858, 2.299999952316284, 4.5])

The above code snippet creates a array holding all the elements which are of same type i.e. float. 'f', designates float or real number.

# NumPy Arrays

But ndarray provides many operations on top of the arrays and hence needs to preferred over the
built-in array module.

### Differences between Python's `array` module and NumPy `ndarray`

Both the Python `array` module and NumPy `ndarray` provide ways to store collections of numbers. However, they have significant differences, especially when it comes to performance and functionality for numerical computing:

1.  **Flexibility and Features:**
    *   **`array.array`**: This is a basic array type that stores a sequence of *fixed-type* elements. It's more memory-efficient than a standard Python list for homogeneous numerical data but offers very limited functionality beyond basic creation and indexing.
    *   **NumPy `ndarray`**: This is the core data structure of the NumPy library. It's a powerful N-dimensional array object designed for high-performance numerical operations. It comes with a vast ecosystem of functions for mathematical operations, linear algebra, Fourier transforms, random number generation, and more.

2.  **Performance:**
    *   **`array.array`**: While more efficient than Python lists for numerical data, its operations are generally performed in Python loops, which can be slower for large datasets.
    *   **NumPy `ndarray`**: NumPy arrays are implemented in C and Fortran, allowing for highly optimized, vectorized operations. This means that operations on entire arrays are executed much faster than explicit Python loops, which is crucial for scientific computing and data analysis.

3.  **Data Types:**
    *   **`array.array`**: Requires a single character type code (e.g., `'i'` for signed int, `'f'` for float) to specify the data type for all elements in the array. This type is fixed upon creation.
    *   **NumPy `ndarray`**: Supports a much wider range of numerical data types (e.g., `int8`, `uint16`, `float32`, `float64`, `complex128`), and allows explicit control over the precision and storage of data.

4.  **Dimensions:**
    *   **`array.array`**: Primarily designed for one-dimensional arrays (vectors).
    *   **NumPy `ndarray`**: Designed for N-dimensional arrays, meaning it can efficiently handle 1D (vectors), 2D (matrices), and higher-dimensional data structures.

5.  **Usage:**
    *   **`array.array`**: Useful when you need a simple, memory-efficient, one-dimensional array of basic types, and don't require advanced mathematical operations.
    *   **NumPy `ndarray`**: The go-to choice for almost all numerical and scientific computing in Python due to its performance, rich functionality, and support for multi-dimensional data.

<b> NumPy array creation from List<b>

In [5]:
import numpy as np

Use np.array to create the ndarrays from Lists

In [6]:
my_list = [1, 2, 3, 4, 5]
my_np_array = np.array(my_list)
print("type   : ", type(my_np_array))
my_np_array

type   :  <class 'numpy.ndarray'>


array([1, 2, 3, 4, 5])

In [7]:
my_float_list = [1.1, 2.2, 3.3]
my_float_array = np.array(my_float_list)
my_float_array

array([1.1, 2.2, 3.3])

All elements should be of same type. If types do not match, Numpy will upcast if possible.

In [8]:
my_mixed_list = [1.2, 1, 2, 3.4]
my_array = np.array(my_mixed_list)
my_array

array([1.2, 1. , 2. , 3.4])

Explicit type also can set for the array.

In [9]:
my_list = [1, 2, 3, 4, 5]
my_array = np.array(my_list, dtype="int32")
my_array

array([1, 2, 3, 4, 5], dtype=int32)

In [10]:
my_float_list = [1.1, 2.2, 3.3]
my_array = np.array(my_list, dtype="float32")
my_array

array([1., 2., 3., 4., 5.], dtype=float32)

Array can be converted back to the list.

In [11]:
my_list = [1, 2, 3, 4, 5]
my_array = np.array(my_list, dtype="int32")

back_to_list = my_array.tolist()
back_to_list

[1, 2, 3, 4, 5]

<b> Array Attributes<b>

In [12]:
x = np.array([1, 2, 3, 4])

In [13]:
#Data type of array
x.dtype

dtype('int64')

In [14]:
#Number of dimensions i.e. ndim
x.ndim

1

In [15]:
#Shape of array i.e. rows , column information
x.shape

(4,)

In [16]:
#Number of elements in array i.e. size = rows * columns
x.size

4

<b> Creating Arrays filled in with default values<b>

In many cases, it will be required to use the arrays which has some default elements present inside it like ones or zeros or some fixed number. NumPy provides built-in functions to create such arrays directly.

In [43]:
#Create array filled of 5 ones
np.ones(5)
# np.ones(5).dtype

array([1., 1., 1., 1., 1.])

In [18]:
#Create array filled of 5 ones which are integers
np.ones(5, dtype="int")

array([1, 1, 1, 1, 1])

In [19]:
#Create array filled of 5 zeros
np.zeros(5)

array([0., 0., 0., 0., 0.])

In [20]:
#Create array filled of 5 zeros which are integers
np.zeros(5, dtype=int)

array([0, 0, 0, 0, 0])

In [21]:
#Create uninitialized array of integers, any garbage value will appear in it
np.empty(5)

array([1.18825e-318, 0.00000e+000, 0.00000e+000, 0.00000e+000,
       0.00000e+000])

<b> Creating linear sequence arrays<b>

In [22]:
#Create an array having element range between 0 to 10 spaced at distance of 1
np.arange(0, 10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [23]:
#Create an array having element range between 0 to 10 spaced at distance of 2
np.arange(0, 10, 2)

array([0, 2, 4, 6, 8])

In [24]:
#Create an array having element range between 5 to 15 spaced at distance of 3
np.arange(5, 15, 3)

array([ 5,  8, 11, 14])

In [25]:
#Create an array having element range between 0 to 1 having 5 values which are equidistant from each other
np.linspace(0, 1, 5)

array([0.  , 0.25, 0.5 , 0.75, 1.  ])

In [26]:
#Create an array having element range between 10 to 20 having 5 values which are equidistant from each other
np.linspace(10, 20, 5)

array([10. , 12.5, 15. , 17.5, 20. ])

### Explanation of `np.linspace(10, 20, 5)`

The `np.linspace(start, stop, num)` function creates `num` evenly spaced samples over the interval `[start, stop]`. In your case, `np.linspace(10, 20, 5)` means:

*   **Start**: 10
*   **Stop**: 20
*   **Number of elements (`num`)**: 5

To find the difference between consecutive elements (the step size), NumPy calculates:


`(stop - start) / (num - 1)`

So, for `(20 - 10) / (5 - 1)`:
`10 / 4 = 2.5`

This means each subsequent element will be 2.5 greater than the previous one:
1.  10
2.  10 + 2.5 = **12.5**
3.  12.5 + 2.5 = 15.0
4.  15.0 + 2.5 = 17.5
5.  17.5 + 2.5 = 20.0

This is why the next element after 10 is 12.5.

<b> Creating multi-dimensional arrays <b>

In [27]:
#Create a two dimensional array with lists as its rows
list1 = [1, 2, 3]
list2 = [4, 5, 6]
list3 = [7, 8, 9]
my_nd_array = np.array([list1, list2, list3])
my_nd_array

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [28]:
#Number of dimensions
my_nd_array.ndim

2

In [29]:
#Shape of array
my_nd_array.shape

(3, 3)

In [30]:
#Size of array  rows * columns
my_nd_array.size

9

In [31]:
#Create a 2-d array with dimensions 3 * 5 filled with all zeros
np.zeros((3,5))

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [32]:
#Create a 2-d array with dimenstions 4 * 4 filled with all ones
np.ones((4,4))

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [33]:
#Create identity matrix of dimensions 4 * 4
np.eye(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

# Exercise

Q1. Create an one dimensional array filled with random numbers. Write a function that prints out the maximum and minimum values from the array

In [34]:
#Try it here :


Q2. Create an one dimensional array filled with random numbers. Write a function that returns an array filled with reciprocal of array elements.

In [35]:
#Try it here :


Q3. Write a function that accepts one dimensional array as input and another number. The funtion should return the array containing all the elements which are greater than the number given as input to the function.

In [36]:
#Try it here :


# Solution

In [37]:
#Solution 1:
def find_min_max(narray):
    min = max = narray[0]

    for value in narray:
        if value > max:
            max = value
        if value < min:
            min = value

    print("Minimum is ", min)
    print("Maximum is ", max)

my_array = np.array([1, 2, 3, 4, 5])
find_min_max(my_array)

Minimum is  1
Maximum is  5


In [38]:
#Solution 2:
def find_reciprocal(narray):
    rec_list = []

    for value in narray:
        rec_list.append(1 / value)

    return np.array(rec_list)

list1 = [1, 2, 3, 4, 5]
rec_array = find_reciprocal(np.array(list1))
print("Original array : ", np.array(list1))
print("Reciprocal array : ", rec_array)

Original array :  [1 2 3 4 5]
Reciprocal array :  [1.         0.5        0.33333333 0.25       0.2       ]


In [39]:
# Q3 Solution :

def find_bigger_numbers(narray, number):
    big_numbers = []

    for value in narray:
        if value > number:
            big_numbers.append(value)

    return np.array(big_numbers)

list1 = [1, 2, 3, 4, 5, 6, 7]
array = np.array(list1)
big_numbers = find_bigger_numbers(array, 3)
big_numbers



array([4, 5, 6, 7])